File: p2p/java/ChimeraP2PEnvelope.java
package io.chimera.p2p;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;

/** Canonical P2P integrity helper; authentication is supplied by the transport. */
public final class ChimeraP2PEnvelope {
    private ChimeraP2PEnvelope() {}

    public static String sha256(String payload) {
        try {
            byte[] digest = MessageDigest.getInstance("SHA-256")
                    .digest(payload.getBytes(StandardCharsets.UTF_8));
            StringBuilder out = new StringBuilder(64);
            for (byte b : digest) out.append(String.format("%02x", b));
            return out.toString();
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }
}

File: p2p/csharp/ChimeraP2PEnvelope.cs
namespace Chimera.P2P;

public sealed record ChimeraP2PEnvelope(
    int Version,
    string Type,
    string NodeId,
    long Sequence,
    string PayloadHash,
    object Payload,
    string[]? Capabilities = null,
    string? Signature = null);

public static class ChimeraP2PIntegrity
{
    public static string Sha256(string text)
    {
        using var sha = System.Security.Cryptography.SHA256.Create();
        var bytes = sha.ComputeHash(System.Text.Encoding.UTF8.GetBytes(text));
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }
}
